File: examples/rag_chain.py
from __future__ import annotations

import argparse

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import RunnablePassthrough
from langchain_openai import ChatOpenAI

from langchain_sourcey import SourceyRetriever


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Answer a question over a published Sourcey docs site."
    )
    parser.add_argument(
        "--site-url",
        default="https://sourcey.com/docs",
        help="Published Sourcey docs root.",
    )
    parser.add_argument(
        "--question",
        default="How does Sourcey document MCP servers?",
        help="Question to answer with retrieved documentation.",
    )
    args = parser.parse_args()

    retriever = SourceyRetriever(site_url=args.site_url, top_k=3)

    prompt = ChatPromptTemplate.from_template(
        """Answer the question using the documentation context below.

{context}

Question: {question}"""
    )

    chain = (
        RunnablePassthrough.assign(context=(lambda x: x["question"]) | retriever)
        | prompt
        | ChatOpenAI(model="gpt-4.1-mini")
        | StrOutputParser()
    )

    print(chain.invoke({"question": args.question}))


if __name__ == "__main__":
    main()

File: js/examples/rag-chain.ts
import { StringOutputParser } from "@langchain/core/output_parsers";
import type { Document } from "@langchain/core/documents";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { RunnablePassthrough, RunnableSequence } from "@langchain/core/runnables";
import { ChatOpenAI } from "@langchain/openai";

import { SourceyRetriever } from "../src/index.js";

const retriever = new SourceyRetriever({
  siteUrl: "https://sourcey.com/docs",
  topK: 3,
});

const prompt = ChatPromptTemplate.fromTemplate(
  `Answer the question using the documentation context below.

{context}

Question: {question}`
);

const formatDocs = (docs: Document[]) =>
  docs.map((doc) => doc.pageContent).join("\n\n");

const chain = RunnableSequence.from([
  {
    context: retriever.pipe(formatDocs),
    question: new RunnablePassthrough(),
  },
  prompt,
  new ChatOpenAI({ model: "gpt-4.1-mini" }),
  new StringOutputParser(),
]);

const answer = await chain.invoke("How does Sourcey document MCP servers?");

console.log(answer);
